flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/KNN.scala

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.classification
+
+import org.apache.flink.api.common.operators.Order
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.scala.DataSetUtils._
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.common._
+import org.apache.flink.ml.math.Vector
+import org.apache.flink.ml.metrics.distances.{DistanceMetric, EuclideanDistanceMetric}
+import org.apache.flink.ml.pipeline.{FitOperation, PredictDataSetOperation, Predictor}
+import org.apache.flink.util.Collector
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/** Implements a k-nearest neighbor join.
+  *
+  * This algorithm calculates `k` nearest neighbor points in training set for each points of
+  * testing set.
+  *
+  * @example
+  * {{{
+  *     val trainingDS: DataSet[Vector] = ...
+  *     val testingDS: DataSet[Vector] = ...
+  *
+  *     val knn = KNN()
+  *       .setK(10)
+  *       .setBlocks(5)
+  *       .setDistanceMetric(EuclideanDistanceMetric())
+  *
+  *     knn.fit(trainingDS)
+  *
+  *     val predictionDS: DataSet[(Vector, Array[Vector])] = knn.predict(testingDS)
+  * }}}
+  *
+  * =Parameters=
+  *
+  * - [[org.apache.flink.ml.classification.KNN.K]]
+  * Sets the K which is the number of selected points as neighbors. (Default value: '''None''')
+  *
+  * - [[org.apache.flink.ml.classification.KNN.Blocks]]
+  * Sets the number of blocks into which the input data will be split. This number should be set
+  * at least to the degree of parallelism. If no value is specified, then the parallelism of the
+  * input [[DataSet]] is used as the number of blocks. (Default value: '''None''')
+  *
+  * - [[org.apache.flink.ml.classification.KNN.DistanceMetric]]
+  * Sets the distance metric to calculate distance between two points. If no metric is specified,
```

Comment:
Sets the distance metric to calculate distance between two points -> Sets the distance metric _we use_ to calculate distance between two points